plugins/gateway/tyk/v3.2.2/plugin.go

```diff
@@ -0,0 +1,190 @@
+
+	reqBody, err := ioutil.ReadAll(req.Body)
+	if err != nil {
+		return nil, fmt.Errorf("failed to read request body: %v", err)
+	}
+
+	resBody, err := ioutil.ReadAll(res.Body)
+	if err != nil {
+		return nil, fmt.Errorf("failed to read response body: %v", err)
+	}
+
+	telemetry := models.Telemetry{
+		DestinationAddress:   ":" + port, // No destination ip for now
+		DestinationNamespace: "",
+		Request: &models.Request{
+			Common: &models.Common{
+				TruncatedBody: false,
+				Body:          strfmt.Base64(reqBody),
+				Headers:       createHeaders(req.Header),
+				Version:       req.Proto,
+			},
+			Host:   host,
+			Method: req.Method,
+			Path:   req.URL.String(),
+		},
+		RequestID: uuid.NewV4().String(),
+		Response: &models.Response{
+			Common: &models.Common{
+				TruncatedBody: false,
+				Body:          strfmt.Base64(resBody),
+				Headers:       createHeaders(res.Header),
+				Version:       req.Proto,
+			},
+			StatusCode: strconv.Itoa(res.StatusCode),
+		},
+		Scheme:        req.Proto,
+		SourceAddress: req.RemoteAddr,
+	}
+
+	return &telemetry, nil
+}
+
+func getHostAndPortFromTargetURL(targetURL string) (host, port string) {
+	parsedHost, err := url.Parse(targetURL)
+	if err != nil {
+		return targetURL, ""
+	}
+	host = parsedHost.Hostname()
+	port = parsedHost.Port()
+	if port == "" {
+		port = "80"
+	}
+
+	return
+}
+
+func createHeaders(headers map[string][]string) []*models.Header {
+	ret := []*models.Header{}
+
+	// TODO support multiple values for a header
+	for header, values := range headers {
+		ret = append(ret, &models.Header{
+			Key:   header,
+			Value: values[0],
+		})
+	}
+	return ret
+}
+
+func newAPIClient(host string) *client.APIClarityPluginsTelemetriesAPI {
+	cfg := client.DefaultTransportConfig()
+	transport := httptransport.New(host, "/api", cfg.Schemes)
+	apiClient := client.New(transport, strfmt.Default)
+	return apiClient
+}
+
+// This is a hack. Currently there is an open bug in Tyk that the APIDefinition is nil
+// https://github.com/TykTechnologies/tyk/issues/3612
+// It does not work in the response phase, so need to propagate this information from a previous phase
+func apiDefinitionRetriever(currentCtx interface{}) *apidef.APIDefinition {
+	contextValues := reflect.ValueOf(currentCtx).Elem()
+	contextKeys := reflect.TypeOf(currentCtx).Elem()
+
+	if contextKeys.Kind() == reflect.Struct {
+		for i := 0; i < contextValues.NumField(); i++ {
+			rv := contextValues.Field(i)
+			reflectValue := reflect.NewAt(rv.Type(), unsafe.Pointer(rv.UnsafeAddr())).Elem().Interface()
+
+			reflectField := contextKeys.Field(i)
+
+			if reflectField.Name == "Context" {
+				apiDefinitionRetriever(reflectValue)
+			} else if fmt.Sprintf("%T", reflectValue) == "*apidef.APIDefinition" {
+				apidefinition := apidef.APIDefinition{}
+				b, _ := json.Marshal(reflectValue)
+				e := json.Unmarshal(b, &apidefinition)
+				if e == nil {
+					return &apidefinition
+				}
+			}
+		}
+	}
+
+	return nil
+}
+
+func main() {}
```

Comment:
`func main()` is it a must?

Reply:
This is what the docs says
https://tyk.io/docs/plugins/supported-languages/golang/#plugin-development-flow